client/src/main/java/io/pravega/client/stream/impl/Pinger.java

```diff
@@ -90,26 +83,26 @@ private void startPeriodicPingTxn() {
         }
     }
 
-    /*
+    /**
      *  Ping all the transactions present in the list. Controller client performs retries in case of a failures. On a
      *  failure the particular transaction it is removed from the ping list.
      */
     private void pingTransactions() {
-        log.info("Start sending transaction pings.");
-        Map<UUID, CompletableFuture<Void>> pingResult =
-                txnList.stream().collect(Collectors.toMap(Function.identity(),
-                        uuid -> {
-                            log.debug("Sending ping request for txn ID: {} with lease: {}", uuid, txnLeaseMillis);
-                            return controller.pingTransaction(stream, uuid, txnLeaseMillis);
-                        }));
-        Futures.allOf(pingResult.values())
-               .whenComplete((v, ex) -> pingResult.entrySet().stream()
-                                                  .filter(e -> !Futures.isSuccessful(e.getValue()))
-                                                  .forEach(e -> {
-                                                      log.warn("Ping Transaction for txn ID:{} failed", e.getKey(),
-                                                              getException(e.getValue()));
-                                                  }));
-        log.trace("Completed sending transaction pings.");
+        try {
+            log.info("Start sending transaction pings.");
+            txnList.stream().forEach(uuid -> {
+                log.debug("Sending ping request for txn ID: {} with lease: {}", uuid, txnLeaseMillis);
+                controller.pingTransaction(stream, uuid, txnLeaseMillis)
```

Comment:
We could wrap these individual pings in retry blocks.

Reply:
Couple of points and questions:

* `controller.pingTransaction(..)` internally has a retry defined at `io.pravega.client.stream.impl.ControllerImpl#retryConfig`. This will take care of all exceptions related to connection drop.

1.  A `StatusRuntimeException` is thrown incase of connection drop which is retried because of `ControllerImpl#retryConfig`.
2. A ping `controller.pingTransaction(...)` on an aborted Txn the completableFuture is completed exceptionally e.g:
```
io.pravega.controller.store.stream.StoreException$DataNotFoundException: /transactions/activeTx/scope/stream/0/00000000-0000-0000-0000-000000000001
```

Is there any other exceptions that can be thrown ? (apart from RetriesExhaustedException.)

* In case of an exception during `controller.commitTransaction(..)` @ `io.pravega.client.stream.impl.EventStreamWriterImpl.TransactionImpl#commit` we should stop pinging, right?

Reply:
> Is there any other exceptions that can be thrown ? (apart from RetriesExhaustedException which can cause the termination of the periodic ping operation )

Periodic ping operation can fail if there is an exception thrown while submitting the futures. 
For example, if `txnList` is null or if `controller`.pingTransaction has controller in null or invalid state and throws exception.

Reply:
I would like to add support for removing aborted transactions from the pinger. But this would require a protocol change on the controller. So I've filed https://github.com/pravega/pravega/issues/2830